web/source/kmwembedded.ts

```diff
@@ -361,14 +362,25 @@ namespace com.keyman.text {
       /* Clear any pending (non-popup) key */
       osk.vkbd.keyPending = null;
 
-      // Changes for Build 353 to resolve KMEI popup key issues      
+      // Changes for Build 353 to resolve KMEI popup key issues
       keyName=keyName.replace('popup-',''); //remove popup prefix if present (unlikely)
 
-      // Can't just split on '-' because some layers like ctrl-shift contain it.
-      let separatorIndex = keyName.lastIndexOf('-');
+      let displayLayer = keymanweb.core.keyboardProcessor.layerId;
 
-      if (separatorIndex > 0) {
-        keyName = keyName.substring(separatorIndex+1);
+      // Regex for 'display layer'-'virtual key name'+'optional functional layer'
+      // Can't just split on '-' because some layers like ctrl-shift contain it.
+      // Virtual key can be T_, U_, K_, or ISO 9995
+      let rx = new RegExp("(.*)-([TKU]_[^+]*|[A-E]\\d\\d)\\+?(.*)?");
+      let matches = keyName.match(rx);
+      if (matches == null) {
+        return false;
+      }
+      displayLayer = matches[1];
+      keyName = matches[2];
+      let functionalLayer = '';
+      if (matches[3]) {
+        functionalLayer = matches[3];
+        keyName += '+' + functionalLayer;
       }
```

Comment:
```suggestion
      keyName = matches[2] + (matches[3] ? '+' + matches[3] : '');
```

Additional optimisation possible: if we include the `+` in the final regex capture group, we can simplify to:

```
      const matches = keyName.match(/^(.+)-([TKU]_[^+]+)\+?(.+)?$/);
      if (matches == null) {
        return false;
      }
      // matches[1]: displayLayer
      // matches[2]: keyId
      // matches[3]: functionalLayer
      keyName = matches[2] + matches[3];
```

Note I am including the extra detail in comments so that it is compiled out to reduce the size of the compiled code -- critical for web.

Reply:
I've incorporated all your comments above in the latest commit.

@keymanapp/test-bot retest